README.md
# Heartcheck::Cas

[![Build Status](https://travis-ci.org/locaweb/heartcheck-cas.svg?branch=master)](https://travis-ci.org/locaweb/heartcheck-cas)
[![Code Climate](https://codeclimate.com/github/locaweb/heartcheck-cas/badges/gpa.svg)](https://codeclimate.com/github/locaweb/heartcheck-cas)

A plugin to check CAS accessibility connection with [heartcheck](https://github.com/locaweb/heartcheck).

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'heartcheck-cas'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install heartcheck-cas

## Usage

You can check any CAS credentials that there's in your app.

```ruby
Heartcheck.setup do |config|
  config.add :cas do |c|
    c.add_service({
        service: 'locaweb',
        username: 'username',
        password: 'password',
        server: 'https://path-to-cas-server/v1/tickets'
    })
  end
end
```

**Check Heartcheck example** [here](https://github.com/locaweb/heartcheck/blob/master/lib/heartcheck/generators/templates/config.rb)

## Development setup using Docker

The Docker Hearthcheck-Cas provides a container with the current stable version of Ruby released and requires you to have these tools available in your local environment:

*   [Docker](https://docs.docker.com/get-docker/)
*   [Docker Compose](https://docs.docker.com/compose/install/)
*   [Bash](https://www.gnu.org/software/bash/)

#### BootStrap Script to run the dockerized environment

 ```bash
 ./scripts/heartcheck-cas setup
 ```

 Run the command `./scripts/heartcheck-cas -h` to see available options.

## Contributing

1.  [Fork it](https://github.com/locaweb/heartcheck-cas/fork)
2.  Create your feature branch ( **git checkout -b my-new-feature** )
3.  Commit your changes ( **git commit -am 'Add some feature'** )
4.  Push to the branch ( **git push origin my-new-feature** )
5.  Create a new Pull Request

## License

*   [MIT License](https://github.com/locaweb/heartcheck-cas/blob/master/LICENSE.txt)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/).

## 1.0.1
### Added
-   Add docker to facilitate development

## 1.0.0
### Changed
-   Add support for ruby 2.7
-   Remove support for ruby < 2.3
-   Change required version for gem heartcheck to fit for newer supported ruby versions

## 0.1.0
### Changed
-   A plugin to check CAS accessibility connection with heartcheck



spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'heartcheck/cas'


spec/heartcheck/cas_spec.rb
require 'spec_helper'

describe Heartcheck::Cas do
  it 'has a version number' do
    expect(Heartcheck::Cas::VERSION).not_to be nil
  end
end


spec/heartcheck/checks/cas_spec.rb
require 'spec_helper'

describe Heartcheck::Checks::Cas do
  let :service do
    {
        service: 'localhost',
        username: 'username',
        password: 'password',
        server: 'https://cas.server.com/v1/tickets'
    }
  end
  
  subject do
    described_class.new.tap do |config|
      config.add_service(service)
    end
  end
  
  describe '#validate' do
    context 'request a ticket to the cas server' do
      it 'error message when raise error' do
        expect(Heartcheck::Cas::HttpClient).
          to receive(:post).and_raise("Yo")

        subject.validate
        expect(subject.instance_variable_get(:@errors)).to contain_exactly("Yo")
      end

      it 'error message with wrong credentials' do
        expect(Heartcheck::Cas::HttpClient).
          to receive(:post).with(service[:server], {
                     'username' => service[:username],
                     'password' => service[:password]
                   }).and_return(double(code: '403'))

        subject.validate

        expect(subject.instance_variable_get(:@errors)).to contain_exactly("Cannot get TGT")
      end
      
      it 'error message with invalid tgt-url' do
        response = double(code: '201')
        expect(response).to receive(:[]).with('Location').and_return('the-tgt-url')
        
        expect(Heartcheck::Cas::HttpClient).to receive(:post).with(service[:server], {
          'username' => service[:username],
          'password' => service[:password]
        }).and_return(response)


        expect(Heartcheck::Cas::HttpClient).to receive(:post).
          with('the-tgt-url', { 'service' => service[:service] }).
            and_return(double(code: '403'))
        
        subject.validate

        expect(subject.instance_variable_get(:@errors)).to contain_exactly("Cannot get ticket")
      end

      it 'request a ticket and works fine' do
        tgt = double(code: '201')
        expect(tgt).to receive(:[]).with('Location') { 'the-tgt-url' }
        
        expect(Heartcheck::Cas::HttpClient).to receive(:post).with(service[:server], {
          'username' => service[:username],
          'password' => service[:password]
        }).and_return(tgt)

        ticket = double(code: '200')
        expect(ticket).to receive(:body) { 'ticketz' }

        expect(Heartcheck::Cas::HttpClient).to receive(:post).
          with('the-tgt-url', {'service' => service[:service]}).
            and_return(ticket)
        
        subject.validate

        expect(subject.instance_variable_get(:@errors)).to be_empty
      end
    end
  end
end


lib/heartcheck/checks/cas.rb
require 'json'

module Heartcheck
  module Checks
    # Check for a cas service
    # Base is set in heartcheck gem
    class Cas < Base
      def validate
        services.each do |service|
          begin
            tgt = tgt_url(service)

            if tgt.nil?
              @errors << "Cannot get TGT"
              next
            end

            if ticket(service, tgt).nil?
              @errors << "Cannot get ticket"
              next
            end
          rescue StandardError => e
            @errors << e.message
          end
        end
      end

      private

      def tgt_url(service)
        response = Heartcheck::Cas::HttpClient.post(service[:server], {
          'username' => service[:username],
          'password' => service[:password]
        })

        response.code.eql?('201') ? response['Location'] : nil
      end

      def ticket(service, tgt)
        response = Heartcheck::Cas::HttpClient.post(tgt, {'service' => service[:service]})
        response.code.eql?('200') ? response.body : nil
      end
    end
  end
end
